bridges/snowbridge/primitives/router/src/outbound/v2.rs

```diff
@@ -0,0 +1,1704 @@
+		let result = instructions.matcher().match_next_inst_while(
+			|_| true,
+			|inst| {
+				return match inst {
+					AliasOrigin(..) => Err(ProcessMessageError::Yield),
+					_ => Ok(ControlFlow::Continue(())),
+				}
+			},
+		);
+		ensure!(result.is_err(), SendError::NotApplicable);
+
+		let mut converter =
+			XcmConverter::<ConvertAssetId, ()>::new(&message, expected_network, agent_id);
+		let message = converter.convert().map_err(|err| {
+			log::error!(target: TARGET, "unroutable due to pattern matching error '{err:?}'.");
+			SendError::Unroutable
+		})?;
+
+		// validate the message
+		let (ticket, _) = OutboundQueue::validate(&message).map_err(|err| {
+			log::error!(target: TARGET, "OutboundQueue validation of message failed. {err:?}");
+			SendError::Unroutable
+		})?;
+
+		Ok(((ticket.encode(), XcmHash::from(message.id)), Assets::default()))
+	}
+
+	fn deliver(blob: (Vec<u8>, XcmHash)) -> Result<XcmHash, SendError> {
+		let ticket: OutboundQueue::Ticket = OutboundQueue::Ticket::decode(&mut blob.0.as_ref())
+			.map_err(|_| {
+				log::trace!(target: TARGET, "undeliverable due to decoding error");
+				SendError::NotApplicable
+			})?;
+
+		let message_id = OutboundQueue::deliver(ticket).map_err(|_| {
+			log::error!(target: TARGET, "OutboundQueue submit of message failed");
+			SendError::Transport("other transport error")
+		})?;
+
+		log::info!(target: TARGET, "message delivered {message_id:#?}.");
+		Ok(message_id.into())
+	}
+}
+
+/// Errors that can be thrown to the pattern matching step.
+#[derive(PartialEq, Debug)]
+pub enum XcmConverterError {
+	UnexpectedEndOfXcm,
+	EndOfXcmMessageExpected,
+	WithdrawAssetExpected,
+	DepositAssetExpected,
+	NoReserveAssets,
+	FilterDoesNotConsumeAllAssets,
+	TooManyAssets,
+	ZeroAssetTransfer,
+	BeneficiaryResolutionFailed,
+	AssetResolutionFailed,
+	InvalidFeeAsset,
+	SetTopicExpected,
+	ReserveAssetDepositedExpected,
+	InvalidAsset,
+	UnexpectedInstruction,
+	TooManyCommands,
+	AliasOriginExpected,
+	InvalidOrigin,
+}
+
+macro_rules! match_expression {
+	($expression:expr, $(|)? $( $pattern:pat_param )|+ $( if $guard: expr )?, $value:expr $(,)?) => {
+		match $expression {
+			$( $pattern )|+ $( if $guard )? => Some($value),
+			_ => None,
+		}
+	};
+}
+
+pub struct XcmConverter<'a, ConvertAssetId, Call> {
+	iter: Peekable<Iter<'a, Instruction<Call>>>,
+	message: Vec<Instruction<Call>>,
+	ethereum_network: NetworkId,
+	agent_id: AgentId,
+	_marker: PhantomData<ConvertAssetId>,
+}
+impl<'a, ConvertAssetId, Call> XcmConverter<'a, ConvertAssetId, Call>
+where
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+{
+	pub fn new(message: &'a Xcm<Call>, ethereum_network: NetworkId, agent_id: AgentId) -> Self {
+		Self {
+			message: message.clone().inner().into(),
+			iter: message.inner().iter().peekable(),
+			ethereum_network,
+			agent_id,
+			_marker: Default::default(),
+		}
+	}
+
+	pub fn convert(&mut self) -> Result<Message, XcmConverterError> {
```

Comment:
For V2, an `InboundMessage` can contain many commands. This conversion code looks like it only matches enough instructions to produce a single command?

Also we need support for XCM::Transact, ie convert it to gateway `CallContract` command.

We should support any combination of `UnlockNativeToken`, `MintForeignToken`, `CallContract` commands.

Reply:
Yeah, will improve it later.